File: 240202/Patrion.cpp
#include <bits/stdc++.h>
using namespace std;
#define rng(i,a,b) for(int i=int(a);i<=int(b);i++)
#define gnr(i,b,a) for(int i=int(b);i>=int(a);i--)
typedef long long ll;

#define N 100010
#define MOD 1000000007

ll datas[N];
ll sum[N];
ll backval[N][19]; // backval[i][j] is number of indicies that idx>=i and datas[idx]==val 
ll DP[N][19][19]; 
ll cnt[N][19];
// DP[i][j][k] is number of (idx1,idx2) pairs, S.T. i<=idx1<idx2 and sum[idx1]==tar+j-k, datas[idx2]==k
// Here, tar is half of total sum of datas.

void Solve(){
    int n;
    cin>>n;
    rng(i,0,n-1){
        cin>>datas[i];
        sum[i]=(i==0?0:sum[i-1])+datas[i];
    }
    if(sum[n-1]%2==1){
        cout<<0<<'\n';
        return;
    }
    ll tar=sum[n-1]/2;
    rng(j,0,18) backval[n][j]=0;
    gnr(i,n-1,0){
        rng(j,0,18){
            backval[i][j]=(i==n-1?0:backval[i+1][j])+(datas[i]==j-9);
        }
    }
    gnr(i,n-1,0){
        rng(j,0,18){
            rng(k,0,18){
                DP[i][j][k]=(i==n-1?0:DP[i+1][j][k])+(sum[i]==tar+j-k?backval[i+1][k]:0);
            }
        }
    }
    
    ll res=0;
    // Case 1 : Array divided outside idx of swaped elements
    rng(i,0,n-1){
        rng(j,0,18) cnt[i][j]=(i==0?0:cnt[i-1][j])+(datas[i]==j-9);
    }
    rng(i,0,n-2){
        if(sum[i]==tar){
            res++;
            res+=(i+1)*(i)/2+(n-i-1)*(n-i-2)/2;
            rng(j,0,18){
                res-=cnt[i][j]*(cnt[i][j]-1)/2;
                res-=(cnt[n-1][j]-cnt[i][j])*(cnt[n-1][j]-cnt[i][j]-1)/2;
            }
        }
    }

    // Case 2 : Array divided between swaped elements
    rng(i,0,n-2){
        rng(val,-9,+9){ // Consider swap datas[i], datas[j](=val) S.T. i<j
            if(datas[i]==val) continue;
            // For all k(k>=i) that satisfy sum[0:k]=tar+datas[i]-val, add the number of j that j>k and val==datas[j]
            res+=DP[i][datas[i]+9][val+9];
        }
    }
    cout<<(res%MOD+MOD)%MOD<<'\n';
}

int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL);
    int t=1;
    cin>>t;
    while(t--){
        Solve();
    }
    return 0;
}

File: 240202/Patrion.py
N = 1010
MOD = 1000000007

datas = [0] * N
sum_arr = [0] * N
backval = [[0] * 19 for _ in range(N)]
DP = [[[0] * 19 for _ in range(19)] for _ in range(N)]
cnt = [[0] * 19 for _ in range(N)]

def solve():
    n = int(input())
    datas = list(map(int, input().split()))
    sum_arr[0] = datas[0]
    for i in range(1, n):
        sum_arr[i] = sum_arr[i - 1] + datas[i]
    
    if sum_arr[n - 1] % 2 == 1:
        print(0)
        return
    
    tar = sum_arr[n - 1] // 2
    
    for j in range(19):
        backval[n][j]=0
    for i in range(n - 1, -1, -1):
        for j in range(19):
            backval[i][j] = (0 if i == n - 1 else backval[i + 1][j]) + (datas[i] == j - 9)
    
    for i in range(n - 1, -1, -1):
        for j in range(19):
            for k in range(19):
                DP[i][j][k] = (0 if i == n - 1 else DP[i + 1][j][k]) + (backval[i + 1][k] if sum_arr[i] == tar + j - k else 0)
    
    res = 0
    
    # Case 1: Array divided outside idx of swapped elements
    for i in range(n - 1):
        if sum_arr[i] == tar:
            res += 1
            res += (i + 1) * i // 2 + (n - i - 1) * (n - i - 2) // 2
            for j in range(19):
                res -= cnt[i][j] * (cnt[i][j] - 1) // 2
                res -= (cnt[n - 1][j] - cnt[i][j]) * (cnt[n - 1][j] - cnt[i][j] - 1) // 2
    
    # Case 2: Array divided between swapped elements
    for i in range(n - 1):
        for val in range(-9, 10):
            if datas[i] == val:
                continue
            res += DP[i][datas[i] + 9][val + 9]
    
    print((res % MOD + MOD) % MOD)

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        solve()